Support OpenThread Border Router in Zephyr

markus-becker-tridonic-com opened the pull request:
* Add UART platform backend for the NCP.
* Remove log_strdup warnings for OpenThread.
* Start CDC_ACM UART earlier so that it is ready in time for OpenThread
  initialisation.
* Tested on FRDM-K64 with RF2XX connected via USB to a Linux machine
  running wpantund.

tbursztyka reviewed (commented):
Could you split into atomic patches? 
- cdc acm change
- log_strdup
- ot uart platform
- ot radio changes
- sample added.

(not specifically in that order)

markus-becker-tridonic-com commented:
Depends on https://github.com/zephyrproject-rtos/openthread/pull/6.

markus-becker-tridonic-com commented:
> Could you split into atomic patches?
> 
> * cdc acm change
> * log_strdup
> * ot uart platform
> * ot radio changes
> * sample added.
> 
> (not specifically in that order)

I could do that, but I wanted to make sure that I have everything in that is required for an OT border router.

finikorg commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -1067,7 +1067,7 @@ static const struct uart_driver_api cdc_acm_driver_api = {
 			    &cdc_acm_init, &cdc_acm_dev_data_##x,	\
 			    &cdc_acm_config_##x,			\
-			    APPLICATION,				\
+			    POST_KERNEL,				\
```
This should be separate patch.

jfischer-phytec-iot replied:
Better separate PR, and probably not necessary, why you can not initialize your application at APPLICATION level?

markus-becker-tridonic-com replied:
https://github.com/zephyrproject-rtos/zephyr/pull/23287

It's not the application, it is the OpenThread networking stack - which gets initialized earlier than the application. The CDC_ACM UART is required to be ready by the time the OpenThread-NCP starts since it immediately uses the UART to inform the Linux OpenThread tools about a (re-)start.

markus-becker-tridonic-com replied:
https://github.com/zephyrproject-rtos/zephyr/pull/23287

markus-becker-tridonic-com replied:
Rebased onto master after https://github.com/zephyrproject-rtos/zephyr/pull/23287 has been merged.

markus-becker-tridonic-com commented:
Any idea why shippable might fail on `test_philosophers_tracing`?

markus-becker-tridonic-com commented:
> Better separate PR, and probably not necessary, why you can not initialize your application at APPLICATION level?

It's not the application, it is the OpenThread networking stack - which gets initialized earlier than the application. The CDC_ACM UART is required to be ready by the time the OpenThread-NCP starts since it immediately uses the UART to inform the Linux OpenThread tools about a (re-)start.

rlubos commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,155 @@
+		__ASSERT_NO_MSG(err == 0);
+	} else {
+		uart_irq_tx_disable(dev);
+		ot_uart->tx_busy = 0;
```
Is `tx_busy` actually used anywhere?
Also, I find it confusing that the same `ot_uart` is used as an argument name and a global variable.

markus-becker-tridonic-com replied:
Should be fixed now. tx_busy and tx_finished are used now.

rlubos commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,155 @@
+
+otError otPlatUartEnable(void)
+{
+	(&ot_uart)->dev = device_get_binding(
```
Why `(&ot_uart)->dev` and not simply `ot_uart.dev`?

markus-becker-tridonic-com replied:
Should be fixed now.

markus-becker-tridonic-com commented on subsys/net/l2/openthread/Kconfig:
```diff
@@ -201,6 +201,59 @@ config OPENTHREAD_DHCP6_SERVER
+    bool "Run SPINEL over USB-ACM"
+    help
+      Is the SPINEL device a USB-CDC-ACM device
+#TODO: enable USB device stack
```
USB support is complete. The TODO is about how to depend on the relevant USB CDC ACM config.

markus-becker-tridonic-com commented on samples/net/sockets/ot_border_router/src/border-router.c:
```diff
@@ -0,0 +1,32 @@
+/* border-router.c - OpenThread NCP Border Router */
```
> * Why is it inside `sockets` directory? It's not releated to sockets whatsoever.
I used a sockets example, copied it and stripped it down.
> * I'm not convinced if `ot_border_router` is a good name for this sample. One could get an impression this sample is a **complete** OT Border Router solution, while the truth is it's merely a Network CoProcessor. A host running wpandtund is still needed in order to have a Border Router.
Yes, it is indeed needed.
> * That said, I'd either call it `ot_ncp` and move it within `samples/net/` or call it simply `ncp` and create a new subdirectory `samples/net/ot/`
I would prefer the latter approach, but the echo samples with ot overlay would make sense in that folder as well. They can however be run without OpenThread, thus it also makes sense to keep in the `sockets/` directory.
How about `samples/net/openthread_ncp/`?

jukkar replied:
> How about `samples/net/openthread_ncp/`?

What about isolating all OpenThread specific stuff to `samples/net/openthread`and create ncp directory under it?

rlubos replied:
+1 for a separate directory, `openthread` sounds sane.

markus-becker-tridonic-com replied:
Should be fixed now.

markus-becker-tridonic-com commented on samples/net/sockets/ot_border_router/src/border-router.c:
```diff
@@ -0,0 +1,32 @@
+{
+	LOG_INF(APP_BANNER, CONFIG_APPLICATION_VERSION);
+
+	while (1) {
```
I think the main thread is a critical thread and if it finished, Zephyr will complain.

markus-becker-tridonic-com commented on samples/net/sockets/ot_border_router/src/border-router.c:
```diff
@@ -0,0 +1,32 @@
+
+#include <net/net_mgmt.h>
+#include <net/net_event.h>
+#include <net/net_conn_mgr.h>
```
Likely not. Left-overs from the socket sample I used.

markus-becker-tridonic-com replied:
Removed.

jukkar commented on samples/net/sockets/ot_border_router/src/border-router.c:
```diff
@@ -0,0 +1,32 @@
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ot_br, LOG_LEVEL_DBG);
```
> For such a simple use case it would probably be enough just to use `printk`

I would rather prefer that logging subsystem is used for all prints.

rlubos commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,155 @@
+{
+	size_t cnt = ring_buf_put((&ot_uart)->tx_ringbuf, aBuf, aBufLength);
+
+	if (atomic_set(&(&ot_uart)->tx_busy, 1) == 0)
```
It's explained in https://docs.zephyrproject.org/latest/contribute/#coding-style

markus-becker-tridonic-com replied:
Should be fixed now.

markus-becker-tridonic-com commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -401,14 +406,23 @@ static int openthread_init(struct net_if *iface)
 	platformShellInit(ot_context->instance);
 #endif
 
+#if defined(CONFIG_OPENTHREAD_NCP)
```
this is in alignment with code above and below.

rlubos replied:
I've pushed a PR which unified the approach: https://github.com/zephyrproject-rtos/zephyr/pull/23298
Please, follow this scheme.

markus-becker-tridonic-com replied:
OK, will wait for https://github.com/zephyrproject-rtos/zephyr/pull/23298 to be merged before changing this.

markus-becker-tridonic-com replied:
Fixed.

markus-becker-tridonic-com commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -344,6 +344,24 @@ otError otPlatRadioEnergyScan(otInstance *aInstance, u8_t aScanChannel,
 }
 
+otError otPlatRadioGetCcaEnergyDetectThreshold(otInstance *aInstance,
+			int8_t *aThreshold)
```
Should be fixed now.

markus-becker-tridonic-com commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -344,6 +344,24 @@ otError otPlatRadioEnergyScan(otInstance *aInstance, u8_t aScanChannel,
+}
+
+otError otPlatRadioSetCcaEnergyDetectThreshold(otInstance *aInstance,
+			int8_t aThreshold)
```
Should be fixed now.

markus-becker-tridonic-com commented on subsys/net/l2/openthread/Kconfig:
```diff
@@ -201,6 +201,59 @@ config OPENTHREAD_DHCP6_SERVER
+    bool "Network Co-Processor as Radio"
+    help
+      Enable NCP in OpenThread stack as radio-only
+    select OPENTHREAD_NCP
```
Should be fixed now.

markus-becker-tridonic-com commented on samples/net/sockets/ot_border_router/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+  min_flash: 140
+sample:
+  description: Runs the OpenThread stack as NCP BR
+  name: OpenThread Border Router
```
Should be fixed now.

markus-becker-tridonic-com commented on samples/net/sockets/ot_border_router/prj.conf:
```diff
@@ -0,0 +1,51 @@
+# Generic networking options
```
The config file has been cleaned-up. Need to check on hardware that it still is working.

markus-becker-tridonic-com replied:
Should be fixed now.

markus-becker-tridonic-com commented on samples/net/sockets/ot_border_router/prj.conf:
```diff
@@ -0,0 +1,51 @@
+CONFIG_NET_UDP=y
+CONFIG_NET_TCP=y
+CONFIG_NET_IPV6=y
+CONFIG_NET_IPV4=y
```
Should be fixed now.

markus-becker-tridonic-com commented on samples/net/sockets/ot_border_router/prj.conf:
```diff
@@ -0,0 +1,51 @@
+# For linux<->qemu connectivity, create a new conf file and swap the
+# addresses (so that peer address is ending to 2).
+CONFIG_NET_CONFIG_SETTINGS=y
+CONFIG_NET_CONFIG_MY_IPV6_ADDR="2001:db8::2"
```
Should be fixed now.

markus-becker-tridonic-com commented on samples/net/sockets/ot_border_router/overlay-ot.conf:
```diff
@@ -0,0 +1,90 @@
+CONFIG_CPLUSPLUS=y
```
Should be fixed now.

markus-becker-tridonic-com commented on samples/net/sockets/ot_border_router/overlay-802154.conf:
```diff
@@ -0,0 +1,18 @@
+CONFIG_BT=n
```
Has been removed.

markus-becker-tridonic-com commented on samples/net/sockets/ot_border_router/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+project(or_border_router)
+
+
+target_sources(                     app PRIVATE src/border-router.c)
```
Should be fixed now.

markus-becker-tridonic-com commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,155 @@
+
+#include "platform-zephyr.h"
+
+const struct openthread_uart {
```
Should be fixed now.

markus-becker-tridonic-com commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,155 @@
+OT_UART_DEFINE(ot_uart, 4096)
+
+enum {
+	kReceiveFifoSize = 128,
```
Should be fixed now.

markus-becker-tridonic-com commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,155 @@
+		.tx_ringbuf = &_name##_tx_ringbuf,			\
+	}
+
+OT_UART_DEFINE(ot_uart, 4096)
```
Should be fixed now.

nandojve commented:
@markus-becker-tridonic-com This is so nice! I was wondering if we can expect a hello-world demo of this on Zephyr? I mean, a place with some docs that shows the steps to put all together including Linux BR, NCP and another board executing ping on any external IP?

I believe it will be very helpfull to gain high adoption.

@nandojve @benpicco

tkons commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,155 @@
+	rd_len = uart_fifo_read(ot_uart->dev,
+				data, kReceiveFifoSize);
+	if (rd_len > 0)
+		otPlatUartReceived(data, rd_len);
```
This is a description how to correctly use otPlatUartReceived and  otPlatUartSendDone:
1) You should add rxfifo, and when in uart_rx_handle function, load bytes from driver's fifo into your uart.c rxfifo. (see usb_shell.c as an example). Then call otSysEventSignalPending.
2) In uart_tx_handle function, call otSysEventSignalPending when TX fifo is empty. (this is not OK for generic driver, but acceptable for NCP) 
3) Add to otSysProcessDrivers function following code:
a) call otPlatUartReceived pointing to your uart.c rxfifo (sometime once, sometime twice, depending if rxfifo looped) . Consume uart.c rxfifo.
b) call otPlatUartSendDone if txfifo empty
4) The "ideal" driver shall support no-interrupt device (but can be added later, as most probably, will never be used)

markus-becker-tridonic-com replied:
@tkons, I did not find any usb_shell.c file within Zephyr. Which file did you mean?

tkons replied:
You are right, the reference driver is shell_uart.c from subsys/shell. Sorry for my mistake

markus-becker-tridonic-com replied:
I have pushed a rework according to this comment to https://github.com/markus-becker-tridonic-com/zephyr/tree/openthread-ncp-border-router-reworked . Testing this on hardware failed up to now. Would be great if you could review, @tkons.

markus-becker-tridonic-com replied:
Should be fixed now.

markus-becker-tridonic-com commented on samples/net/sockets/ot_border_router/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+common:
+  harness: net
+  tags: net socket
```
Should be fixed now.

markus-becker-tridonic-com commented:
> @markus-becker-tridonic-com This is so nice! I was wondering if we can expect a hello-world demo of this on Zephyr? I mean, a place with some docs that shows the steps to put all together including Linux BR, NCP and another board executing ping on any external IP?
> 
> I believe it will be very helpfull to gain high adoption.
> 
> @nandojve @benpicco

@nandojve, I don't have full instructions yet.

It boils down to:
1. Build the BR uC image:
`west build -b tri_n4m_br -d build-br -s zephyr/samples/net/openthread/ncp -- -DBOARD_ROOT=app/ -DCONF_FILE="prj.conf overlay-tri-n4m-br.conf"`
2. Build OpenThread for your Linux system according to upstream instructions https://openthread.io/guides/border-router/build to get wpanctl and wpantund.
3. Configure wpantund for the ttyACM device
4. Build an OpenThread device, e.g. the echo_client with overlay-ot.conf

Obviously we are doing this on our hardware, so YMMV.

nandojve commented:
Hi @markus-becker-tridonic-com, I enabled SAMV71 + RF2XX as leader this weekend. My test was conducted using Debian 10 and NCP over USB. I had some issues with otbr-web, However, after manually formation I was able to see status on web.

I need invest more time with Openthread to have a full thread enabled OTBR but NCP is working.

I hope soon you can merge this on Zephyr.

markus-becker-tridonic-com commented:
> Hi @markus-becker-tridonic-com, I enabled SAMV71 + RF2XX as leader this weekend. My test was conducted using Debian 10 and NCP over USB. I had some issues with otbr-web, However, after manually formation I was able to see status on web.
> 
> I need invest more time with Openthread to have a full thread enabled OTBR but NCP is working.
> 
> I hope soon you can merge this on Zephyr.

I have done a re-work to address the review comments (see https://github.com/zephyrproject-rtos/zephyr/pull/23249#discussion_r390439317), but that is not yet functional.

markus-becker-tridonic-com commented:
I have fixed the rework and force-pushed to this branch/PR and tested on our device successfully.

There is a small fix required:
`
diff --git a/drivers/ieee802154/ieee802154_rf2xx.c b/drivers/ieee802154/ieee802154_rf2xx.c
index f2d81b3725..fd1c4e9ced 100644
--- a/drivers/ieee802154/ieee802154_rf2xx.c
+++ b/drivers/ieee802154/ieee802154_rf2xx.c
@@ -448,16 +448,16 @@ static int rf2xx_filter(struct device *dev,
 #if defined(CONFIG_NET_L2_OPENTHREAD)
 static void rf2xx_handle_ack(struct rf2xx_context *ctx, u8_t seq_number)
 {
-	rf2xx_ack_psdu[0] = ACK_FRAME_TYPE;
+	rf2xx_ack_psdu[0] = RF2XX_ACK_FRAME_TYPE;
 	rf2xx_ack_psdu[2] = seq_number;
 
 	if (ctx->trx_trac == RF2XX_TRX_PHY_STATE_TRAC_SUCCESS_DATA_PENDING) {
-		rf2xx_ack_psdu[0] |=  ACK_FRAME_PENDING_BIT;
+		rf2xx_ack_psdu[0] |= RF2XX_ACK_FRAME_PENDING_BIT;
 	}
 
 	rf2xx_ack_frame.data = rf2xx_ack_psdu;
 
-	if (ieee802154_radio_handle_ack(ctx->iface, rf2xx_ack_pkt) != NET_OK) {
+	if (ieee802154_radio_handle_ack(ctx->iface, &rf2xx_ack_pkt) != NET_OK) {
 		LOG_INF("ACK packet not handled.");
 	}
 }
`
@nandojve would you apply that or shall I make it part of this PR?

markus-becker-tridonic-com commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,155 @@
+		ot_uart->tx_busy = 0;
+	}
+
+	otPlatUartSendDone();
```
Should be fixed now.

rlubos commented:
@tkons Could you revisit the uart part?

markus-becker-tridonic-com reviewed (commented):
> @tkons Could you revisit the uart part?

I think I did. In the last force-push the comments from @tkons and you should be addressed.

rlubos commented:
@markus-becker-tridonic-com Sure thanks, I will re-review. I just wanted @tkons to have a look at the uart changes to confirm he's good with them. We all want to push this PR forward :)

markus-becker-tridonic-com commented:
I am still fixing a few of the checkpatch warnings/errors.

markus-becker-tridonic-com commented:
@rlubos are you able to check this on Nordic SoC?

rlubos commented:
@markus-becker-tridonic-com Yes, I will.

nandojve commented:
> I have fixed the rework and force-pushed to this branch/PR and tested on our device successfully.
> 
> There is a small fix required:
> 
> ```
> diff --git a/drivers/ieee802154/ieee802154_rf2xx.c b/drivers/ieee802154/ieee802154_rf2xx.c
> index f2d81b3725..fd1c4e9ced 100644
> --- a/drivers/ieee802154/ieee802154_rf2xx.c
> +++ b/drivers/ieee802154/ieee802154_rf2xx.c
> @@ -448,16 +448,16 @@ static int rf2xx_filter(struct device *dev,
>  #if defined(CONFIG_NET_L2_OPENTHREAD)
>  static void rf2xx_handle_ack(struct rf2xx_context *ctx, u8_t seq_number)
>  {
> -	rf2xx_ack_psdu[0] = ACK_FRAME_TYPE;
> +	rf2xx_ack_psdu[0] = RF2XX_ACK_FRAME_TYPE;
>  	rf2xx_ack_psdu[2] = seq_number;
>  
>  	if (ctx->trx_trac == RF2XX_TRX_PHY_STATE_TRAC_SUCCESS_DATA_PENDING) {
> -		rf2xx_ack_psdu[0] |=  ACK_FRAME_PENDING_BIT;
> +		rf2xx_ack_psdu[0] |= RF2XX_ACK_FRAME_PENDING_BIT;
>  	}
>  
>  	rf2xx_ack_frame.data = rf2xx_ack_psdu;
>  
> -	if (ieee802154_radio_handle_ack(ctx->iface, rf2xx_ack_pkt) != NET_OK) {
> +	if (ieee802154_radio_handle_ack(ctx->iface, &rf2xx_ack_pkt) != NET_OK) {
>  		LOG_INF("ACK packet not handled.");
>  	}
>  }
> ```
> 
> @nandojve would you apply that or shall I make it part of this PR?
> 
> @rlubos, @tkons shall i resolve the comments or how do i request a re-review?

I detected same here, please, feel free to send PR!

markus-becker-tridonic-com commented:
> I detected same here, please, feel free to send PR!

https://github.com/zephyrproject-rtos/zephyr/pull/23525

rlubos commented on samples/net/openthread/ncp/prj.conf:
```diff
@@ -0,0 +1,109 @@
+# Network shell
+# CONFIG_NET_SHELL=y
+
+# ???
```
I think we don't need `net_config` or shell.

markus-becker-tridonic-com replied:
removed.

rlubos commented on samples/net/openthread/ncp/prj.conf:
```diff
@@ -0,0 +1,109 @@
+CONFIG_NET_CONFIG_NEED_IPV6=y
+CONFIG_NET_CONFIG_NEED_IPV4=n
+
+CONFIG_NET_SOCKETS=y
```
Sockets should not be needed either.

markus-becker-tridonic-com replied:
removed.

rlubos commented on samples/net/openthread/ncp/src/border-router.c:
```diff
@@ -0,0 +1,23 @@
+/* border-router.c - OpenThread NCP Border Router */
```
Could we call the file `main.c`? I've already expressed my concerns regarding calling the sample a border router.

Also, I believe we need some documentation for the sample. At minimum, we should link to the official OpenThread documentation (https://openthread.io/guides/border-router) and emphasize that this sample is the NCP part of the whole solution.

markus-becker-tridonic-com replied:
documentation has been added.

border-router.c has been renamed to main.c.

rlubos commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,220 @@
+static void uart_tx_handle(/*struct openthread_uart *ot_uart*/)
+{
+	u32_t len;
+
```
Extra empty line.

markus-becker-tridonic-com replied:
fixed

rlubos replied:
Doesn't seem so.

rlubos commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,220 @@
+
+	/* Process UART RX */
+	while (
+				(
```
This indentation does not seem right.

markus-becker-tridonic-com replied:
fixed.

rlubos replied:
Not fixed as well. Did you miss a push?

Just make it something like:
```
	while ((len = ring_buf_get_claim(ot_uart.rx_ringbuf,
					 (u8_t **)&data,
					 ot_uart.rx_ringbuf->size) > 0) {
```

rlubos commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,220 @@
+
+#define RX_FIFO_SIZE 128
+
+static void uart_rx_handle(/*const struct openthread_uart *ot_uart*/)
```
`static void uart_rx_handle(void)`, let's not leave a dead code.

markus-becker-tridonic-com replied:
fixed.

rlubos commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,220 @@
+	}
+}
+
+static void uart_tx_handle(/*struct openthread_uart *ot_uart*/)
```
Same here, dead code.

markus-becker-tridonic-com replied:
fixed.

rlubos commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,220 @@
+
+	/* Wait 1 sec for the host to do all settings */
+	LOG_INF("Wait for host to settle");
+	k_busy_wait(1000000);
```
Is the `k_busy_wait` really needed here, couldn't we let other tasks work in the meantime (as it would happen with `k_sleep`)?

markus-becker-tridonic-com replied:
This is the same as in https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/subsys/usb/cdc_acm/src/main.c#L119.

jukkar replied:
Even if it is like that in other sample, it does not mean it is correct thing to do. I would like to understand why we should have `k_busy_wait()` here instead of `k_sleep()`. Have you tried to use the k_sleep(), does it cause issues?

markus-becker-tridonic-com replied:
Tested just now on hardware with `k_busy_wait()` wpantund starts up just fine. With `k_sleep()` wpantund goes into 
```
wpan0 => [
        "NCP:State" => "uninitialized:fault"
```

Not an expert in the USB stack. So, I would keep it like this and someone knowledgeable from the USB team to take a look into this question here and in the CDC-ACM sample.

jukkar replied:
Ok, that is good info. Could you add a comment to the code about your findings so we do not need wonder later why the busy wait is being used there.

markus-becker-tridonic-com replied:
Done.

jfischer-phytec-iot replied:
If k_sleep() does not work then there is probably a bug or issue, anyway I suggest to change it to `k_busy_wait(USEC_PER_SEC * 1);`

markus-becker-tridonic-com replied:
I did not notice the different time-base between `k_busy_wait(us)` and `k_sleep(ms)`. Once moving to a millisecond sleep, it also works with `k_sleep()`. Fixed and rebased.

rlubos commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,220 @@
+#endif
+
+	uart_irq_callback_user_data_set(ot_uart.dev,
+		uart_callback, (void *)&ot_uart);
```
Either move the entire argument list to a new line, or fix the indentation.

markus-becker-tridonic-com replied:
fixed.

rlubos reviewed (commented):
Thanks, it's getting into shape.

rlubos commented:
Also, I think @tbursztyka made a good point about splitting the changes into separate commits.

markus-becker-tridonic-com committed:
net: openthread: Remove log_strdup warnings for OpenThread

Log messages from the OpenThread logging system were not log_strdup()

Signed-off-by: Markus Becker <markus.becker@tridonic.com>

markus-becker-tridonic-com committed:
net: openthread: Add UART platform backend for NCP

Currently based on CDC-ACM. Can possibly be used with plain UART as
well.

Signed-off-by: Markus Becker <markus.becker@tridonic.com>

markus-becker-tridonic-com committed:
net: openthread: Add NCP start-up and required platform changes

When NCP starts-up some of the initialisation functions of a regular
OpenThread device do not need to be called, because they get triggered
by wpantund via UART. Instead NCP initialisation needs to be called.

A small typo has been fixed as well.

Also initialisation for raw link packet interface has been added. Can be
used for picking up 802.15.4 frames and interpreting them in the
application.

Signed-off-by: Markus Becker <markus.becker@tridonic.com>

markus-becker-tridonic-com committed:
net: openthread: Add NCP sample

OpenThread NCP sample application.

Tested on FRDM-K64 with RF2XX connected via USB to a Linux machine
running wpantund.

Signed-off-by: Markus Becker <markus.becker@tridonic.com>

markus-becker-tridonic-com commented:
> Also, I think @tbursztyka made a good point about splitting the changes into separate commits.

> cdc acm change
has been merged already, not part of this PR anymore.

> log_strdup
> ot uart platform
> ot radio changes
> sample added
done.

markus-becker-tridonic-com commented:
OK, seems gitlint is happy now. The PR currently also includes the commit from https://github.com/zephyrproject-rtos/zephyr/pull/23525 since I need that to successfully compile.

markus-becker-tridonic-com commented:
@rlubos I think all comments except https://github.com/zephyrproject-rtos/zephyr/pull/23249#discussion_r393731583 have been addressed.

rlubos reviewed (commented):
Thanks for all the fixes, the PR looks pretty good now. We still miss a `west.yml` update though, pointing to a newer OT revision (see https://github.com/zephyrproject-rtos/zephyr/pull/23554 to get what I mean). The commit with a manifest update should point to your PR in OT (`pull/6/head`).

In meantime, I'll try to test the changes on nRF hardware.

jfischer-phytec-iot commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -22,6 +22,7 @@ LOG_MODULE_REGISTER(net_l2_openthread, CONFIG_OPENTHREAD_L2_LOG_LEVEL);
 #include <openthread/cli.h>
```
>Also initialisation for raw link packet interface has been added. Can be
used for picking up ZigBee frames and interpreting them in the
application.

Where is it? What is a _ZigBee frame_? Is OpenThread not based on IEEE802.15.4?

markus-becker-tridonic-com replied:
It is part of upstream openthread API. using it here:
https://github.com/zephyrproject-rtos/zephyr/pull/23249/files#diff-bf3f4a47fa194d5e0de369c2cfe68f1aR214-R217
https://github.com/zephyrproject-rtos/zephyr/pull/23249/files#diff-b1484649f09207a98f1953e0ef947964R413-R414

The mapping to https://github.com/zephyrproject-rtos/openthread/blob/zephyr/etc/cmake/options.cmake#L124-L127 is done in https://github.com/zephyrproject-rtos/openthread/pull/6/files#diff-83aeef700a5a192d68e562d9ccfe3725R36-R38.

Using the API you can get access to all IEEE802.15.4, also non-Thread packets like ZigBee packets.

jfischer-phytec-iot replied:
Than please fix your commit and replace ZigBee with IEEE802.15.4.

markus-becker-tridonic-com replied:
Replaced ZigBee with 802.15.4 in the commit message. Anything else I should change?

rlubos commented on subsys/net/l2/openthread/openthread.c:
```diff
@@ -401,13 +406,25 @@ static int openthread_init(struct net_if *iface)
+	}
+
+	if (IS_ENABLED(CONFIG_OPENTHREAD_RAW)) {
+		otLinkRawSetEnable(ot_context->instance, true);
```
This line throws a warning. You need to include the following header:
`#include <openthread/link_raw.h>`

markus-becker-tridonic-com replied:
Added.

rlubos commented on samples/net/openthread/ncp/overlay-tri-n4m-br.conf:
```diff
@@ -0,0 +1,35 @@
+# TRI n4m BR
```
Could you elaborate a bit what is this device? Is this some kind of shield? I've tried to google it, but quick search didn't provide anything useful.

I'm asking, because from user perspective I'd like to know **when** should this overlay be applied, what purpose does it serve. As far as I understand you run FRDM-K64F with 802.15.4 radio attached, but there's no information whatsoever in the documentation on what configuration is supported by this overlay.

markus-becker-tridonic-com replied:
It is this device https://www.tridonic.com/com/en/products/net4more-borderROUTER-PoE-Thread.asp, also listed here: https://www.threadgroup.org/What-is-Thread#certifiedproducts.

The device internally uses a Linux CPU, NXP K64F (not the FRDM board), and AT86RF233.
It could be used to also use for the FRDM board with an ATREB233-XPRO with modifications. The overlay might serve as an example for this.

Did you need an overlay for nrf52840?

rlubos replied:
> Did you need an overlay for nrf52840?

No, once I sorted out the compilation errors (as mentioned [here](https://github.com/zephyrproject-rtos/zephyr/pull/23249#discussion_r394352574)) the NCP worked in the default configuration (with docker image of BR).

I think it's worth to mention this configuration in the documentation then, see how it's done in `echo_client` for instance:
https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/net/sockets/echo_client#building-and-running
Sadly not all overlays are documented in there, but it gives an overview of what's supported.

Some samples specify also a board-specific configs (https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/net/sockets/echo_client/boards), but if it's a board+shield combo I'm not sure if it's any useful.

markus-becker-tridonic-com replied:
I would add the following to README.rst.

```
There are configuration files for different boards and setups in the
echo-client directory:

- :file:`prj.conf`
  Generic config file, normally you should use this.

- :file:`overlay-tri-n4m-br.conf`
  This is an overlay for the dedicated Thread Border Router hardware 
  https://www.tridonic.com/com/en/products/net4more-borderROUTER-PoE-Thread.asp.
  The board support is not part of the Zephyr repositories, but the
  product is based on NXP K64 and AT86RF233. This file can be used as an
  example for a development set-up based on development boards.
```

Do you want to add a sentence on NRF?

rlubos replied:
Just fix the first sentence, it's not `echo-client`.

I don't think you need to explicitly mention nRF here, I believe it's safe to assume that all OT/15.4 samples work on nRF (at least that's my goal).

markus-becker-tridonic-com replied:
done and rebased.

rlubos commented on subsys/net/l2/openthread/Kconfig:
```diff
@@ -201,6 +201,67 @@ config OPENTHREAD_DHCP6_SERVER
+	  Enable border router support in OpenThread stack
+
+config OPENTHREAD_NCP
+	bool "Network Co-Processor"
```
`uart.c` unconditionally depends on ring buffer and interrupt driver UART api, I suggest adding the following here:
```
select RING_BUFFER
select UART_INTERRUPT_DRIVEN
```
Otherwise, the NCP sample does not build in basic configuration (w/o overlay).

markus-becker-tridonic-com replied:
Added.

rlubos reviewed (commented):
I left some more feedback from my tryouts on nRF52840. The good news is that after some tweaks in the configuration (mentioned in the comment) I've successfully ran border router with Zephyr nRF52840 NCP.

jfischer-phytec-iot commented on samples/net/openthread/ncp/overlay-tri-n4m-br.conf:
```diff
@@ -0,0 +1,35 @@
+CONFIG_SERIAL=y
+CONFIG_UART_INTERRUPT_DRIVEN=y
+CONFIG_UART_LINE_CTRL=y
+CONFIG_CDC_ACM_IAD=y
```
CDC_ACM_IAD is enabled per default and IMO not necessary on linux host

markus-becker-tridonic-com replied:
removed. this is not a linux host though. it is for an NXP K64 MCU.

rlubos commented:
Thanks for the updates, I have only one last comment (I'm pretty sure I've already mentioned it).

These changes are tightly coupled with your PR in OT repository: https://github.com/zephyrproject-rtos/openthread/pull/6/files. You need to update `west.yml` file in this PR, in order to point to the updated OT version.

For starters, you should add a commit that sets OpenThread revision to `pull/6/head`. After the OpenThread PR is merged, you should update the commit to point to the actual SHA on the `zephyr` branch.

markus-becker-tridonic-com committed:
net: openthread: Use updated OpenThread module

The relevant upstream changes in OpenThread for NCP support have
been integrated in
https://github.com/zephyrproject-rtos/openthread/pull/6.

Signed-off-by: Markus Becker <markus.becker@tridonic.com>

markus-becker-tridonic-com commented:
> Thanks for the updates, I have only one last comment (I'm pretty sure I've already mentioned it).
> 
> These changes are tightly coupled with your PR in OT repository: https://github.com/zephyrproject-rtos/openthread/pull/6/files. You need to update `west.yml` file in this PR, in order to point to the updated OT version.
> 
> For starters, you should add a commit that sets OpenThread revision to `pull/6/head`. After the OpenThread PR is merged, you should update the commit to point to the actual SHA on the `zephyr` branch.

OK, it wasn't totally clear, that I was supposed to do that or that it is done in the process.

Done.

rlubos commented:
@markus-becker-tridonic-com Please also check https://github.com/zephyrproject-rtos/openthread/pull/6, I've asked for a rebase there. 

https://github.com/zephyrproject-rtos/openthread/pull/6/commits/e4927d973c22101a211f63633f3721965b15f515 is no longer needed, as it was taken from upstream.

markus-becker-tridonic-com commented:
> @markus-becker-tridonic-com Please also check [zephyrproject-rtos/openthread#6](https://github.com/zephyrproject-rtos/openthread/pull/6), I've asked for a rebase there.
> 
> [zephyrproject-rtos/openthread@e4927d9](https://github.com/zephyrproject-rtos/openthread/commit/e4927d973c22101a211f63633f3721965b15f515) is no longer needed, as it was taken from upstream.

did so. rebases onto origin/master once more.

rlubos reviewed (approved):
Thank you. Adding DNM until dependent OT PR is merged.

jfischer-phytec-iot commented on subsys/net/l2/openthread/Kconfig:
```diff
@@ -201,6 +201,69 @@ config OPENTHREAD_DHCP6_SERVER
+	bool "Run SPINEL over USB-ACM"
+	help
+	  Is the SPINEL device a USB-CDC-ACM device
+	depends on OPENTHREAD_NCP && USB_CDC_ACM
```
Usually the `help` is the last item, please fix it.

markus-becker-tridonic-com replied:
Fixed.

jfischer-phytec-iot commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,217 @@
+		return OT_ERROR_FAILED;
+	}
+
+	/* Wait 1 sec for the host to do all settings */
```
This comment is not necessary, two line below describe it self clearly.

markus-becker-tridonic-com replied:
Fixed.

jfischer-phytec-iot commented on samples/net/openthread/ncp/src/main.c:
```diff
@@ -0,0 +1,23 @@
+	LOG_INF(APP_BANNER, CONFIG_APPLICATION_VERSION);
+
+	while (1) {
+		k_sleep(MSEC_PER_SEC * 4U);
```
Why not just exit main?

markus-becker-tridonic-com replied:
See discussion in https://github.com/zephyrproject-rtos/zephyr/pull/23249#discussion_r387699909. I checked on hardware and it works. Also checked in code, it seems the main thread is marked non-essential after it has ended.

Fixed.

rlubos commented:
@markus-becker-tridonic-com https://github.com/zephyrproject-rtos/openthread/pull/6 is now merged, please update SHA in the manifest to `bc267a62237dac68d1a7443acadd66d0415d155d`

rlubos commented:
SHA is correct now, I've lifted DNM. Seems that you need a rebase though (conflict).

jukkar commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,216 @@
+			uart_tx_handle();
+		}
+	}
+};
```
Unnecessary ; here

markus-becker-tridonic-com replied:
fixed.

jukkar commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,216 @@
+
+static void uart_callback(void *user_data)
+{
+	while (uart_irq_update(ot_uart.dev) &&
```
Please add `ARG_UNUSED(user_data);` as user_data is not used.

jukkar commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,216 @@
+			(u8_t **)&data,
+			ot_uart.rx_ringbuf->size)) > 0) {
+		otPlatUartReceived(data, len);
+		int err = ring_buf_get_finish(
```
Variables should be declared at the beginning of the block.

markus-becker-tridonic-com replied:
fixed.

jukkar commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,216 @@
+	LOG_INF("Wait for host to settle");
+	k_sleep(MSEC_PER_SEC * 1);
+
+	u32_t baudrate = 0U;
```
same here, please introduce baudrate after the ret var

jukkar commented on subsys/net/lib/openthread/platform/uart.c:
```diff
@@ -0,0 +1,216 @@
+#endif
+
+	uart_irq_callback_user_data_set(
+		ot_uart.dev,
```
Indentation, it looks like at least some of the variables could be put previous line

markus-becker-tridonic-com replied:
fixed.

jukkar commented on samples/net/openthread/ncp/Kconfig:
```diff
@@ -0,0 +1,15 @@
+
+menu "Application configuration"
+
+config APPLICATION_VERSION
```
We have Kconfig options in other network samples, we could follow similar naming convention for these. So please add prefix `NET_SAMPLE_...` here so this could be then `config NET_SAMPLE_APPLICATION_VERSION` or `NET_SAMPLE_APP_VERSION`.

markus-becker-tridonic-com replied:
fixed.

jukkar reviewed (changes requested):
Thanks, looks good, just couple of minor nits.

jukkar commented:
@markus-becker-tridonic-com can you check the merge conflict?

markus-becker-tridonic-com commented:
> @markus-becker-tridonic-com can you check the merge conflict?

done
